pkg/zcapld/middleware.go

```diff
@@ -0,0 +1,287 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package zcapld
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"strings"
+
+	httpsig "github.com/igor-pavlenko/httpsignatures-go"
+)
+
+const (
+	// CapabilityInvocationHTTPHeader is the HTTP header expected on zcap'ed HTTP requests.
+	CapabilityInvocationHTTPHeader = "capability-invocation"
+	signatureHeader                = "signature"
+	capabilityParam                = "capability"
+	actionParam                    = "action"
+	keyIDParam                     = "keyId"
+)
+
+// HTTPSigAuthConfig configures the HTTP auth handler.
+type HTTPSigAuthConfig struct {
+	CapabilityResolver CapabilityResolver
+	KeyResolver        KeyResolver
+	VerifierOptions    []VerificationOption
+	Secrets            httpsig.Secrets
+	ErrConsumer        func(error)
```

Comment:
Is there any reason why we have that `ErrConsumer` instead of just a simple logger? It looks more complicated.

Reply:
@aholovko It's so that the application layer can inject their own logger (or otherwise do whatever they want to with the error). Applications can for instance send the log message to some remote storage, or simply configure their own preferred format.

This is a generic HTTP REST handler, I don't think it should have an opinion on how to log messages.